File: python/src/wgrok/allowlist.py
"""Domain/email pattern matching for wgrok allowlist."""

import logging

logger = logging.getLogger(__name__)


class Allowlist:
    """Validates email addresses against a list of allowed patterns.

    Supported pattern formats:
    - "domain.tld" -> matches *@domain.tld (bare domain)
    - "*@domain.tld" -> wildcard prefix match
    - "user@domain.tld" -> exact match (case-insensitive)

    Dangerous patterns with [, ], or ? are rejected.
    """

    def __init__(self, patterns: list[str]) -> None:
        self._patterns: list[tuple[str, str]] = []  # (type, pattern) pairs
        for p in patterns:
            p = p.strip()
            if not p:
                continue
            # Reject patterns with dangerous characters
            if "[" in p or "]" in p or "?" in p:
                logger.warning(f"Rejecting dangerous allowlist pattern: {p}")
                continue
            # Normalize and store with type
            if p.startswith("*@"):
                # Wildcard prefix: *@domain.tld
                domain = p[2:]  # Extract domain part
                self._patterns.append(("wildcard_prefix", domain.lower()))
            elif "@" in p:
                # Exact match: user@domain.tld
                self._patterns.append(("exact", p.lower()))
            else:
                # Bare domain: domain.tld (equivalent to *@domain.tld)
                self._patterns.append(("bare_domain", p.lower()))

    def is_allowed(self, email: str) -> bool:
        """Check if an email address matches any allowed pattern (case-insensitive)."""
        email_lower = email.lower()
        for pattern_type, pattern_value in self._patterns:
            if pattern_type == "exact" and email_lower == pattern_value:
                return True
            if pattern_type in ("wildcard_prefix", "bare_domain") and email_lower.endswith(f"@{pattern_value}"):
                return True
        return False

File: go/allowlist.go
package wgrok

import (
	"log"
	"strings"
)

// patternType represents the classification of an allowlist pattern.
type patternType int

const (
	exact patternType = iota
	wildcardPrefix
	bareDomain
)

// pattern represents a normalized allowlist pattern with its type.
type pattern struct {
	pType patternType
	value string
}

// Allowlist validates email addresses against a list of allowed patterns.
//
// Supported formats:
//   - "domain.tld"      → matches *@domain.tld (bare domain)
//   - "*@domain.tld"    → wildcard prefix match
//   - "user@domain.tld" → exact match (case-insensitive)
//
// Patterns with [, ], or ? are rejected (dangerous).
type Allowlist struct {
	patterns []pattern
}

// NewAllowlist creates an Allowlist from the given patterns.
func NewAllowlist(patterns []string) *Allowlist {
	var normalized []pattern
	for _, p := range patterns {
		p = strings.TrimSpace(p)
		if p == "" {
			continue
		}
		// Reject patterns with dangerous characters
		if strings.ContainsAny(p, "[]?") {
			log.Printf("Rejecting dangerous allowlist pattern: %s", p)
			continue
		}
		pLower := strings.ToLower(p)
		if strings.HasPrefix(pLower, "*@") {
			// Wildcard prefix: *@domain.tld
			domain := pLower[2:]
			normalized = append(normalized, pattern{pType: wildcardPrefix, value: domain})
		} else if strings.Contains(pLower, "@") {
			// Exact match: user@domain.tld
			normalized = append(normalized, pattern{pType: exact, value: pLower})
		} else {
			// Bare domain: domain.tld
			normalized = append(normalized, pattern{pType: bareDomain, value: pLower})
		}
	}
	return &Allowlist{patterns: normalized}
}

// IsAllowed checks if an email address matches any allowed pattern (case-insensitive).
func (a *Allowlist) IsAllowed(email string) bool {
	emailLower := strings.ToLower(email)
	for _, p := range a.patterns {
		switch p.pType {
		case exact:
			if emailLower == p.value {
				return true
			}
		case wildcardPrefix:
			// Check if email is at the domain
			if strings.HasSuffix(emailLower, "@"+p.value) {
				return true
			}
		case bareDomain:
			// Check if email's domain matches
			if strings.HasSuffix(emailLower, "@"+p.value) {
				return true
			}
		}
	}
	return false
}
